spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/inbound/ReactiveMongoDbMessageSource.java

```diff
@@ -0,0 +1,217 @@
+				"'mongoConverter' can not be set when instance was constructed with ReactiveMongoTemplate");
+		this.mongoConverter = mongoConverter;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "mongo:reactive-inbound-channel-adapter";
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext =
+				ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+		TypeLocator typeLocator = this.evaluationContext.getTypeLocator();
+		if (typeLocator instanceof StandardTypeLocator) {
+			//Register MongoDB query API package so FQCN can be avoided in query-expression.
+			((StandardTypeLocator) typeLocator).registerImport("org.springframework.data.mongodb.core.query");
+		}
+		if (this.reactiveMongoTemplate == null) {
+			this.reactiveMongoTemplate = new ReactiveMongoTemplate(this.reactiveMongoDatabaseFactory, this.mongoConverter);
+		}
+		this.initialized = true;
+	}
+
+	/**
+	 * Will execute a {@link Query} returning its results as the Message payload.
+	 * The payload can be either {@link Flux} of {@link Mono} of objects of type
+	 * identified by {{@link #entityClass}, or a single element of type identified by {{@link #entityClass}
+	 * based on the value of {{@link #expectSingleResult} attribute which defaults to 'false' resulting
+	 * {@link org.springframework.messaging.Message} with payload of type
+	 * {@link Flux}. The collection name used in the
+	 * query will be provided in the {@link MongoHeaders#COLLECTION_NAME} header.
+	 */
+	@Override
+	public Object doReceive() {
+		Assert.isTrue(this.initialized, "This class is not yet initialized. Invoke its afterPropertiesSet() method");
+		AbstractIntegrationMessageBuilder<Object> messageBuilder = null;
+		Object value = this.queryExpression.getValue(this.evaluationContext);
+		Assert.notNull(value, "'queryExpression' must not evaluate to null");
+		Query query;
+		if (value instanceof String) {
+			query = new BasicQuery((String) value);
+		}
+		else if (value instanceof Query) {
+			query = ((Query) value);
+		}
+		else {
+			throw new IllegalStateException("'queryExpression' must evaluate to String " +
+					"or org.springframework.data.mongodb.core.query.Query");
+		}
+
+		Assert.notNull(query, "'queryExpression' must not evaluate to null");
+		String collectionName = this.collectionNameExpression.getValue(this.evaluationContext, String.class);
+		Assert.notNull(collectionName, "'collectionNameExpression' must not evaluate to null");
+
+		Object result;
+		if (this.expectSingleResult) {
+			result = this.reactiveMongoTemplate.
+					findOne(query, this.entityClass, collectionName);
+		}
+		else {
+			result = this.reactiveMongoTemplate.
+					find(query, this.entityClass, collectionName);
+		}
+		if (result != null) {
+			messageBuilder = this.getMessageBuilderFactory().withPayload(result)
+					.setHeader(MongoHeaders.COLLECTION_NAME, collectionName);
+		}
+
+//	   TODO: Do we need to add any reactive transaction support?
```

Comment:
I don't think so. This one is only about `read` operation, so no reason in transactions at all.
Plus the real execution is going to happen somewhere downstream on demand.
Therefore any storing in the reactive context does not make sense.